Client validation (#49)
quicwg/version-negotiation: draft-ietf-quic-version-negotiation.md

    organization: Mozilla
    email: ekr@rtfm.com

> normative:

Reviewer: Out of curiosity, does a syntax like `{{Section 6 of !RFC8999}}` not work?

Author: It didn't when I last touched that code.

The server starts incompatible version negotiation by sending a Version
Negotiation packet. This packet SHALL include each entry from the server's set
of Offered Versions (see {{server-fleet}}) in a Supported Version field. The
server MAY add reserved versions (as defined in {{Section 6.3 of QUIC}}) in
Supported Version fields.

> A client MUST ignore a Version Negotiation packet if it contains the original

Reviewer: I'm not a huge fan of repeating normative statements from a normative reference. How about something non-normative that's backed up by the reference:
```suggestion
Clients will ignore a Version Negotiation packet if it contains the original
```

the connection was using QUIC version 1, that connection closure MUST use a
transport error of type `VERSION_NEGOTIATION_ERROR`.

> The client MUST validate the server `Other Versions` field by confirming that it

Reviewer: This requirement only applies to clients that have reacted to a VN packet right? What validation MUST a client do if this connection didn't involve a VN packet? (in other words, why drop the "If a client has reacted to a Version Negotiation packet," part?)

Reviewer 2: Why shouldn't the client not do the same validation during compatible version negotiation?

Author: There are ways in which the two are different, so I think David is right here.

Let's say that there are two groups of versions, the versions in each group compatible with each other, but incompatible with the versions in the other group.  The client prefers a version in the first group, but the client also knows that the versions in the first group are not widely implemented.  As the client doesn't like the added latency of version negotiation, it chooses the second group, which contains a version that is totally OK.  However, if it gets a Version Negotiation packet as a result, it will happily switch to the first group.

So yeah, let's go with the "if VN" condition.

would have attempted the same version with this knowledge of the versions the
server supports. That is, the client would have selected the same version if it

> received Version Negotiation packet that listed the versions in the server's

Reviewer: ```suggestion
received a Version Negotiation packet that listed the versions in the server's
```

the connection was using QUIC version 1, that connection closure MUST use a
transport error of type `VERSION_NEGOTIATION_ERROR`.

> If the client received and acted on a Version Negotiation packet, it client MUST

Reviewer: ```suggestion
If the client received and acted on a Version Negotiation packet, the client MUST
```

> validate the server `Other Versions` field.  The `Other Versions` field is

Reviewer: ```suggestion
validate the server's `Other Versions` field.  The `Other Versions` field is
```
